src/main/scala/org/camunda/dmn/parser/DmnParser.scala

```diff
@@ -153,24 +153,19 @@ class DmnParser(
 
     bkms.exists(bkm =>
       hasDependencyCycle(
-        visit = List(bkm.getId),
+        visit = bkm.getId,
         visited = Set.empty,
         dependencies = dependencies
       ))
   }
 
-  @tailrec
-  private def hasDependencyCycle(visit: List[String],
+  private def hasDependencyCycle(visit: String,
                         visited: Set[String],
                         dependencies: Map[String, Iterable[String]]): Boolean = {
     visit match {
-      case Nil => false
-      case element :: _ if visited.contains(element) => true
-      case element :: tail => hasDependencyCycle(
-        visit = tail ++ dependencies.getOrElse(element, Nil),
-        visited = visited + element,
-        dependencies = dependencies
-      )
+      case element if visited.contains(element) => true
+      case element => dependencies.getOrElse(element, Nil).exists(dependency =>
+        hasDependencyCycle(dependency, visited + element, dependencies))
```

Comment:
ℹ️ I kept this as pattern matching. This could also be switched to a simple if-else:

```java
    if (visited.contains(visit)) {
       true
     } else {
       dependencies.getOrElse(visit, Nil).exists(dependency =>
         hasDependencyCycle(dependency, visited + visit, dependencies)
       )
     }
```

This is similar to what we've had before [it got broken during a refactoring](https://github.com/camunda/dmn-scala/pull/207/commits/3161483aebfb5aef90e01fd32a82fe54e231aa3f#diff-36c33ab259b225915e1f3e920b87002f85d9c6a4bc5e97c0475003b19daf1344L165). Let me know what you prefer!

Reply:
Let's switch to the if-else block. :+1: 
The pattern matching doesn't bring value in this form. Unfortunately, no tail recursion anymore. But it is hard. :laughing: 

:wrench: We could use named parameters for the recursive call.

```scala
    if (visited.contains(visit)) {
       true
     } else {
       dependencies.getOrElse(visit, Nil).exists(dependency =>
         hasDependencyCycle(
           visit = dependency, 
           visited = visited + visit, 
           dependencies = dependencies
         )
       )
     }
```